components/dashboard/src/components/start-workspace.tsx

```diff
@@ -322,16 +356,49 @@ export class StartWorkspace extends React.Component<StartWorkspaceProps, StartWo
         return window.top !== window.self;
     }
 
-    render() {
-        const errorCode = this.state && this.state.errorCode;
-
+    protected renderError() {
+        const { errorCode, errorData } = this.state;
         const startErrorRenderer = this.props.startErrorRenderer;
         if (startErrorRenderer && errorCode) {
             const rendered = startErrorRenderer(errorCode, this.props.service, () => this.startWorkspace(this.props.workspaceId, true, false));
             if (rendered) {
                 return rendered;
             }
         }
+        if (errorCode === ErrorCodes.SETUP_REQUIRED) {
+            return <ApplicationFrame />;
+        }
+        if (errorCode === ErrorCodes.USER_TERMS_ACCEPTANCE_REQUIRED) {
+            return <ApplicationFrame />;
+        }
+        if (errorCode === ErrorCodes.NOT_AUTHENTICATED) {
+            if (errorData?.host && errorData?.scopes && errorData?.messageHint) {
+                return (
+                    <ApplicationFrame service={this.props.service}>
+                        <ShowUnauthorizedError data={errorData} />
+                    </ApplicationFrame>
+                );
+            }
+        }
+        if (errorCode === ErrorCodes.PLAN_DOES_NOT_ALLOW_PRIVATE_REPOS) {
```

Comment:
This is an IO concern which is handled in the `startErrorRenderer`.

Reply:
removing...